src/function/VaryingCDF.cpp

```diff
@@ -0,0 +1,187 @@
+#include <VaryingCDF.h>
+#include <Exceptions.h>
+#include <ContainerTools.hpp>
+#include <ContainerParameter.h>
+#include <ContainerTools.hpp>
+#include <Formatter.hpp>
+#include <Rand.h>
+#include <PDF.h>
+#include <JumpPDF.h>
+
+using ContainerTools::ToString;
+using ContainerTools::GetKeys;
+using ContainerTools::GetValues;
+
+/////////////////////////
+// Constructory Things //
+/////////////////////////
+
+VaryingCDF::VaryingCDF(const std::string& name_ = ""){
+    if(name_=="")
+        fName=std::string("VaryingCDF");
+    else
+        fName=name_;
+}
+VaryingCDF::~VaryingCDF(){
+    delete fCdf;
+    for(std::map<std::string,Function*>::const_iterator parameter_ = fFunctions.begin(); parameter_ !=fFunctions.end(); ++parameter_) {
+        delete parameter_->second;
+    }
+}
+
+VaryingCDF::VaryingCDF(const VaryingCDF& copy_){
+    fFunctions = copy_.fFunctions;
+    fCdf = copy_.fCdf->Clone();
+    fName = copy_.fName;
+}
+
+VaryingCDF&
+VaryingCDF::operator=(const VaryingCDF& copy_){
+    fFunctions = copy_.fFunctions;
+    fCdf = copy_.fCdf->Clone();
+    fName = copy_.fName;
+    return *this;
+}
+
+ConditionalPDF*
+VaryingCDF::Clone() const{
+    return static_cast<ConditionalPDF*> (new VaryingCDF(*this));
+}
+
+ParameterDict
+VaryingCDF::SetupParameters(const std::vector<double>& x2_) const {
+    ParameterDict parameters;
+    for(std::map<std::string,Function*>::const_iterator parameter_ = fFunctions.begin(); parameter_ !=fFunctions.end(); ++parameter_) {
+        parameters[parameter_->first] = (parameter_->second)->operator()(x2_);
+    }
+    return parameters;
+}
+
+double
+VaryingCDF::ConditionalProbability(const std::vector<double>& x_, 
+                                   const std::vector<double>& x2_){
+    if (!fCdf)
+        throw NULLPointerAccessError(Formatter()<<"VaryingCDF:: fCdf is not pointing to a PDF.");
+    fCdf->SetParameters(SetupParameters(x2_));
+    return fCdf->ConditionalProbability(x_,x2_);
+}
+
+void
+VaryingCDF::SetDependance(const std::string& paraName_, const Function* func_){
+    fFunctions[paraName_] = func_->Clone();
+}
+
+double
+VaryingCDF::Integral(const std::vector<double>& mins_, 
+                              const std::vector<double>& maxs_,
+                              const std::vector<double>& x2_) const{
+    if (!fCdf)
+        throw NULLPointerAccessError(Formatter()<<"VaryingCDF:: fCdf is not pointing to a ConditionalPDF.");
+    fCdf->SetParameters(SetupParameters(x2_));
+    return fCdf->Integral(mins_,maxs_,x2_);
+}
+
+void 
+VaryingCDF::SetKernel(ConditionalPDF* PDF_){
+    fCdf=PDF_->Clone();
+}
+
+void 
+VaryingCDF::SetKernel(PDF* PDF_){
+    // BL : If we are passed a PDF we transform to a JumpPDF in order to perserve
+    // the structure of shifting a function to a bin center and integrating over
+    // over bins relative to that shift.
+    fCdf=static_cast<ConditionalPDF*>(new JumpPDF("kernel",PDF_));
```

Comment:
Well the comment says it all really but might not make much sense. So I cast the PDF_ as a JumpPDF so that don't have to worry about shifting the kernel about, then I just set all of the parameters given a bin centre and call the JumpPDF methods (Sample etc). Does that make sense?